### PR TITLE
Support for externalTrafficPolicy=Local in NodePort BPF

### PR DESCRIPTION
**Pull request by gandro**
This adds new `TrafficPolicy` and `NodeName` fields to Cilium's service representation and implements the "Local" traffic policy. When the Kubernetes `externalTrafficPolicy` is set to `Local`, only backends where the node name matches the current node are selected and added to the BPF lbmap. This check of the node name to determine local backends is the same mechanism that is also used by kube-proxy.

When a service has `TrafficPolicy=Local`, the Cilium API only shows and accepts local backends for that service.

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9764)

### CODE REVIEW

**Comment by gandro** [comment]
test-focus K8sServicesTest

**Comment by gandro** [comment]
test-focus K8sServicesTest*

**Comment by gandro** [comment]
test-focus K8sServicesTest*

**Comment by gandro** [comment]
test-me-please

File: daemon/loadbalancer.go [hunk 1]
```diff
@@ -70,13 +70,21 @@ func (h *putServiceID) Handle(params PutServiceIDParams) middleware.Responder {
 		svcType = loadbalancer.SVCTypeClusterIP
 	}
 
+	var svcTrafficPolicy loadbalancer.SVCTrafficPolicy
+	switch params.Config.Flags.TrafficPolicy {
+	case models.ServiceSpecFlagsTrafficPolicyLocal:
+		svcTrafficPolicy = loadbalancer.SVCTrafficPolicyLocal
+	default:
+		svcTrafficPolicy = loadbalancer.SVCTrafficPolicyCluster
```
Comment (brb) [inline daemon/loadbalancer.go:78]: Yep, it makes sense to default to the `Cluster` policy.

File: pkg/service/service.go [hunk 2]
```diff
@@ -68,14 +70,20 @@ func (svc *svcInfo) deepCopyToLBSVC() *lb.SVC {
...
+		Backends:      backends,
+		Type:          svc.svcType,
+		TrafficPolicy: svc.svcTrafficPolicy,
+		Name:          svc.svcName,
+		Namespace:     svc.svcNamespace,
 	}
 }
 
+func (svc *svcInfo) requireNodeLocalBackends() bool {
+	return (svc.svcType == lb.SVCTypeNodePort || svc.svcType == lb.SVCTypeLoadBalancer) &&
```
Comment (brb) [inline pkg/service/service.go]: Shouldn't we add `lb.SVCTypeExternalIPs` to the list? Check https://msazure.club/kubernetes-services-and-iptables/ (a source of the truth when it comes to the kube-proxy semantics).

File: test/k8sT/Services.go [hunk 3]
```diff
@@ -294,17 +294,21 @@ var _ = Describe("K8sServicesTest", func() {
 			}
 		}
 
-		It("Tests NodePort (kube-proxy)", func() {
-			testNodePort(false)
-		})
-
-		It("Tests NodePort (kube-proxy) with externalTrafficPolicy=Local", func() {
+		testExternalTrafficPolicyLocal := func() {
```
Comment (brb) [inline test/k8sT/Services.go:297]: Could you extend the test to check whether a request sent to `k8s2` has been dropped? Can be done as a follow-up.

**Review by brb — CHANGES_REQUESTED** [review]
Amazing work! Just one comment and one nit.

File: pkg/service/service.go (hunk 2 above)
Comment (gandro) [reply pkg/service/service.go]: Good catch. I was following https://github.com/kubernetes/kubernetes/blob/42fe74cd2c2d09a5fa80976bb82c91caa778738e/pkg/api/service/util.go#L71 but forgot that Cilium has additional types that K8s doesn't have. Will fix!

**Comment by gandro** [comment]
test-me-please

File: test/k8sT/Services.go (hunk 3 above)
Comment (gandro) [reply test/k8sT/Services.go:297]: Will do in a separate PR!

**Review by brb — APPROVED** [review]

**Comment by gandro** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
running the tests since connectivity to 1.1.1.1 failed

**Comment by aanm** [comment]
test-me-please

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
Connection to `world` failed again.

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
test-missed-k8s

**Pull request merged**